bin/auto_snapshot.rb

```diff
@@ -1,11 +1,54 @@
 #!/usr/bin/env ruby
 
 require 'cucloud'
+require 'optparse'
 
-num_days = ARGV[0] || 5
+default_days = 5
+
+options = {
+  add_tags: [],
+  num_days: nil,
+  preserve_tags: []
+}
+
+OptionParser.new do |opts|
+  opts.on('--apply-tag Key=Foo,Value=Bar',
+          String,
+          'Appy tag to snapshots created.') do |item|
+    match = /^Key=([^,]+),Value=(.+)$/.match(item)
+    unless match && match[1] && match[2] && !match[3]
+      raise "Argument to --add-tag should be the string 'Key=<key>,Value=<value>'"
+    end
+    options[:add_tags] << { key: match[1], value: match[2] }
+  end
+  opts.on('--num-days N',
+          Integer,
+          "Snapshot volumes without snapshot in past N days (default #{default_days}).") do |item|
+    options[:num_days] = item
+  end
+  opts.on('--preserve-tags x,y,z',
+          Array,
+          'Array of tag keys to preserve from volume.') do |item|
+    item.each do |listitm|
+      options[:preserve_tags] << listitm
+    end
+  end
+  opts.on_tail('-h',
+               '--help',
+               'Command help') do
+    puts opts
+    exit
+  end
+end.parse!
+
+# Restore prior version behavior where you could simply provide the number of days
+# as the sole/first argument.
+options[:num_days] = ARGV[0] || 5 if options[:num_days].nil?
```

Comment:
this line is a little hard to read with the multiple conditionals.  If you use default days in your hash setup at the top this can be simplified.

`options[:num_days] = ARGV[0] if ARGV.length == 1`